src/Dofs/sparsity_pattern.jl

```diff
@@ -1,83 +1,396 @@
+###########################
+# AbstractSparsityPattern #
+###########################
+
+"""
+    AbstractSparsityPattern
+
+Supertype for sparsity pattern implementations, e.g. [`SparsityPattern`](@ref) and
+[`BlockSparsityPattern`](@ref).
+"""
+abstract type AbstractSparsityPattern end
+
+"""
+    n_rows(sp::AbstractSparsityPattern)
+
+Return the number of rows in the sparsity pattern `sp`.
+"""
+n_rows(sp::AbstractSparsityPattern)
+
+"""
+    n_cols(sp::AbstractSparsityPattern)
+
+Return the number of columns in the sparsity pattern `sp`.
+"""
+n_cols(sp::AbstractSparsityPattern)
+
+"""
+    add_entry!(sp::AbstractSparsityPattern, row::Int, col::Int)
+
+Add an entry to the sparsity pattern `sp` at row `row` and column `col`.
+"""
+add_entry!(sp::AbstractSparsityPattern, row::Int, col::Int)
+
+# This is necessary to avoid warning about not importing Base.eachrow when
+# adding docstring before the definitions further down.
+function eachrow end
```

Comment:
Is there a reason why we don't dispatch Base.eachrow?

Reply:
It's not really the same concept; `Base.eachrow` is for matrices and gives the full row, this just give the stored entries of the row. We can call it something else inside of Ferrite though.

Reply:
If the concept is different, then maybe we should rename the function? `row_iterator` maybe?